File: framework/backend/python/packages/k1s0-cache/src/k1s0_cache/patterns/write_behind.py
"""Write-behind (write-back) caching pattern."""

from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field

from k1s0_cache.operations import CacheOperations


@dataclass
class WriteBehindStats:
    """Statistics for write-behind operations."""

    total_writes: int = 0
    total_flushes: int = 0
    total_failures: int = 0


class WriteBehind:
    """Buffers writes and flushes them to a backing store in batches.

    The cache is updated immediately, while the backing store receives
    batched writes at a configurable interval.

    Args:
        cache: Cache operations instance.
        writer: An async callable that persists a batch of key-value pairs.
        batch_size: Maximum batch size per flush.
        flush_interval: Seconds between automatic flushes.
        max_retries: Maximum retry attempts for a failed flush.
    """

    def __init__(
        self,
        cache: CacheOperations,
        writer: Callable[[list[tuple[str, str]]], Awaitable[None]],
        batch_size: int = 100,
        flush_interval: float = 5.0,
        max_retries: int = 3,
    ) -> None:
        self._cache = cache
        self._writer = writer
        self._batch_size = batch_size
        self._flush_interval = flush_interval
        self._max_retries = max_retries
        self._buffer: list[tuple[str, str]] = []
        self._task: asyncio.Task[None] | None = None
        self.stats = WriteBehindStats()

    async def write(self, key: str, value: str) -> None:
        """Write to the cache immediately and buffer for the backing store.

        Args:
            key: The cache key.
            value: The value to write.
        """
        await self._cache.set(key, value)
        self._buffer.append((key, value))
        self.stats.total_writes += 1

    async def flush(self) -> None:
        """Flush the current buffer to the backing store."""
        if not self._buffer:
            return

        batch = self._buffer[: self._batch_size]
        self._buffer = self._buffer[self._batch_size :]

        for attempt in range(self._max_retries):
            try:
                await self._writer(batch)
                self.stats.total_flushes += 1
                return
            except Exception:
                if attempt == self._max_retries - 1:
                    self.stats.total_failures += 1
                    raise

    async def _run(self) -> None:
        """Background loop that flushes periodically."""
        while True:
            await asyncio.sleep(self._flush_interval)
            try:
                await self.flush()
            except Exception:
                pass  # stats already updated in flush

    async def start(self) -> None:
        """Start the background flushing loop."""
        if self._task is None:
            self._task = asyncio.create_task(self._run())

    async def stop(self) -> None:
        """Stop the background flushing loop and perform a final flush."""
        if self._task is not None:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
            self._task = None

        # Final flush of remaining items
        while self._buffer:
            try:
                await self.flush()
            except Exception:
                break

File: framework/backend/csharp/src/K1s0.Cache/Patterns/WriteBehind.cs
using System.Collections.Concurrent;

namespace K1s0.Cache.Patterns;

/// <summary>
/// Implements the write-behind (write-back) pattern: writes are buffered in memory
/// and flushed to the backing store in the background.
/// </summary>
public sealed class WriteBehind : IAsyncDisposable
{
    private readonly ICacheOperations _cache;
    private readonly Func<IReadOnlyList<KeyValuePair<string, string>>, Task> _flushHandler;
    private readonly TimeSpan? _defaultTtl;
    private readonly ConcurrentQueue<KeyValuePair<string, string>> _buffer = new();
    private readonly TimeSpan _flushInterval;
    private CancellationTokenSource? _cts;
    private Task? _backgroundTask;

    private long _totalWrites;
    private long _totalFlushes;
    private long _totalFailures;

    /// <summary>
    /// Initializes a new instance of the <see cref="WriteBehind"/> class.
    /// </summary>
    /// <param name="cache">The cache operations instance.</param>
    /// <param name="flushHandler">A delegate that persists a batch of key-value pairs to the backing store.</param>
    /// <param name="flushInterval">The interval between background flush cycles.</param>
    /// <param name="defaultTtl">Optional default TTL for cached entries.</param>
    public WriteBehind(
        ICacheOperations cache,
        Func<IReadOnlyList<KeyValuePair<string, string>>, Task> flushHandler,
        TimeSpan flushInterval,
        TimeSpan? defaultTtl = null)
    {
        _cache = cache ?? throw new ArgumentNullException(nameof(cache));
        _flushHandler = flushHandler ?? throw new ArgumentNullException(nameof(flushHandler));
        _flushInterval = flushInterval;
        _defaultTtl = defaultTtl;
    }

    /// <summary>
    /// Gets the total number of write operations enqueued.
    /// </summary>
    public long TotalWrites => Interlocked.Read(ref _totalWrites);

    /// <summary>
    /// Gets the total number of successful flush operations.
    /// </summary>
    public long TotalFlushes => Interlocked.Read(ref _totalFlushes);

    /// <summary>
    /// Gets the total number of failed flush operations.
    /// </summary>
    public long TotalFailures => Interlocked.Read(ref _totalFailures);

    /// <summary>
    /// Writes a value to the cache immediately and enqueues it for background persistence.
    /// </summary>
    /// <param name="key">The cache key.</param>
    /// <param name="value">The value to write.</param>
    /// <param name="ttl">Optional TTL override.</param>
    /// <param name="ct">A cancellation token.</param>
    public async Task WriteAsync(string key, string value, TimeSpan? ttl = null, CancellationToken ct = default)
    {
        // Write to cache immediately.
        await _cache.SetAsync(key, value, ttl ?? _defaultTtl, ct).ConfigureAwait(false);

        // Buffer for background flush.
        _buffer.Enqueue(new KeyValuePair<string, string>(key, value));
        Interlocked.Increment(ref _totalWrites);
    }

    /// <summary>
    /// Flushes all buffered writes to the backing store immediately.
    /// </summary>
    public async Task FlushAsync()
    {
        var batch = new List<KeyValuePair<string, string>>();
        while (_buffer.TryDequeue(out var item))
        {
            batch.Add(item);
        }

        if (batch.Count == 0)
        {
            return;
        }

        try
        {
            await _flushHandler(batch).ConfigureAwait(false);
            Interlocked.Increment(ref _totalFlushes);
        }
        catch
        {
            Interlocked.Increment(ref _totalFailures);

            // Re-enqueue items so they are not lost.
            foreach (var item in batch)
            {
                _buffer.Enqueue(item);
            }

            throw;
        }
    }

    /// <summary>
    /// Starts the background flush loop.
    /// </summary>
    public void StartAsync()
    {
        if (_cts is not null)
        {
            return;
        }

        _cts = new CancellationTokenSource();
        _backgroundTask = RunFlushLoopAsync(_cts.Token);
    }

    /// <summary>
    /// Stops the background flush loop and performs a final flush.
    /// </summary>
    public async Task StopAsync()
    {
        if (_cts is null)
        {
            return;
        }

        await _cts.CancelAsync().ConfigureAwait(false);

        if (_backgroundTask is not null)
        {
            try
            {
                await _backgroundTask.ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                // Expected on cancellation.
            }
        }

        // Final flush of remaining items.
        await FlushAsync().ConfigureAwait(false);

        _cts.Dispose();
        _cts = null;
        _backgroundTask = null;
    }

    /// <inheritdoc/>
    public async ValueTask DisposeAsync()
    {
        await StopAsync().ConfigureAwait(false);
    }

    private async Task RunFlushLoopAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                await Task.Delay(_flushInterval, ct).ConfigureAwait(false);
                await FlushAsync().ConfigureAwait(false);
            }
            catch (OperationCanceledException) when (ct.IsCancellationRequested)
            {
                break;
            }
            catch
            {
                // Errors are tracked via TotalFailures; continue the loop.
            }
        }
    }
}
